README.md
# User Management API CLI Tool

A simple command-line interface (CLI) client to the
[Adobe User Management API](https://adobe-apiplatform.github.io/umapi-documentation/en/).

## Table of Contents

* [Feature Overview](#feature-overview)
* [Installation](#installation)
* [Usage Overview](#usage-overview)
* [Configuring](#configuring)
* [Commands](#commands)
* [Appendix: Building the Tool](#appendix-building-the-tool)
* [Getting Help](#getting-help)

# Feature Overview

**User Operations**

* [Query a Single User](#user-read)
* [Query All Users](#user-read-all)
* [Create Single User](#user-create)
* [Create Users in Bulk](#user-create-bulk)
* [Update a User](#user-update)
* [Update Users in Bulk](#user-update-bulk)
* [Remove or Delete a User](#user-delete)
* [Remove or Delete Users in Bulk](#user-delete-bulk)

**Group Operations**

* [Query a Single Group](#group-read)
* [Query All Groups](#group-read-all)
* [Create Single Group](#group-create)
* [Create Groups in Bulk](#group-create-bulk)
* [Update a Group](#group-update)
* [Update Groups in Bulk](#group-update-bulk)
* [Delete a Group](#group-delete)
* [Delete Groups in Bulk](#group-delete-bulk)

"bulk" and "all" operations support multiple input/output formats.

"Read All" output formats:

* Human-readable
* CSV
* [JSONL](http://jsonlines.org)

Bulk create/update/delete formats:

* CSV
* [JSONL](http://jsonlines.org)

# Installation

The recommended method for installing the tool is
[pipx](https://pypa.github.io/pipx/). It not only installs the tool in an
isolated environment, but also makes it available on the shell's `$PATH`
variable.

You can alternatively download the tool as a self-contained executable for
Windows or macOS. Check the [releases
page](https://github.com/adobe/umapi-cli/releases/) for the latest stable
release. Source distributions and wheel builds are also published on each new
release.

## Option 1 - `pipx`

On a system with [pipx](https://pypa.github.io/pipx/installation/) installed,
installing the UMAPI CLI Tool is simple:

``` sh
$ pipx install umapi-cli
```

This puts the `umapi` executable in your shell's `$PATH` and makes it available
from any directory.

``` sh
$ umapi --help
```

## Option 2 - Executable

A self-contained executable is also available for Windows and macOS. The
executable embeds a Python environment, so Python does not need to be installed
on the system. Note that this build is slower to start when executing the tool.

Executables can be found on the [latest
releases](https://github.com/adobe/umapi-cli/releases/latest) page.

# Usage Overview

The tool operates on a series of commands. Each command performs a certain
operation - read single user, read all users, create user/group, etc.

```
$ umapi --help
Usage: umapi [OPTIONS] COMMAND [ARGS]...

Options:
  --env PATH  Path to .env file (optional)
  -t, --test  Run command in test mode
  -v          Enable verbose logging
  -h, --help  Show this message and exit.
  --version   Show the version and exit.

Commands:
  group-create       Create a single user group.
  group-create-bulk  Create groups in bulk from an input file
  group-delete       Delete a single user group
  group-delete-bulk  Delete groups in bulk from input file
  group-read         Get details for a single user group
  group-read-all     Get details for all groups in a console
  group-update       Update information/memberships for a single group
  group-update-bulk  Update groups in bulk from input file
  user-create        Create a single user.
  user-create-bulk   Create users in bulk from an input file
  user-delete        Delete a single user (from org and/or identity...
  user-delete-bulk   Delete users in bulk from input file (from org...
  user-read          Get details for a single user
  user-read-all      Get details for all users belonging to a console
  user-update        Update user information for a single user
  user-update-bulk   Update users in bulk from input file
```

## General Options

The following options apply to any command. They should be specified before the
command being invoked.

For example, to invoke test mode when creating a list of users from a CSV file:

```
$ umapi -t user-create-bulk -f csv -i users.csv
```

The following general options apply to any UMAPI command:

* `--env` - Specify an `env` file containing UMAPI configuration. If this isn't
  specified, then the tool will look for a `.env` file in the current working
  directory. If that isn't present then it will look for the `UMAPI_*`
  environment variables mentioned above, which define config options for the
  UMAPI connection.
* `-t/--test` - Invoke test mode for all UMAPI calls made when running the
  command. Test mode performs all API actions in a "dry run" mode which will
  return results as if the actions were executed, but not perform any actual
  changes to users or groups in the Admin Console. This can also be used for
  user and group read operations, but it does not affect the output of the read
  operation.
* `-v` - Enable logging and control verbosity. Pass `-v` to enable `info` level
  logging. Pass `-vv` to log with more information at the `debug` level. If
  neither option is passed, then the tool will only pass output for errors and
  the results of read operations. **Note:** the tool logs output to stdout.
  Redirect stdout to a file to capture log information.

# Configuring

The CLI tool requires a valid connection to the User Management API. This must
be set up in the [Adobe Developer Console](https://developer.adobe.com/console)
prior to using the tool.

Refer to [this
guide](https://developer.adobe.com/developer-console/docs/guides/authentication/ServerToServerAuthentication/implementation/)
if you need help creating the API integration and credentials. Once setup is
complete, you need three items from the credential page:

* `client_id` - Unique identifier to authorize the API client
* `client_secret` - Secret token to authenticate the connection
* `org_id` - Unique organization identifier

The UMAPI CLI Tool expects these items to be set in the following evironment variables.

* `UMAPI_CLIENT_ID`
* `UMAPI_CLIENT_SECRET`
* `UMAPI_ORG_ID`

These can either be set as system variables (e.g. `export
UMAPI_CLIENT_ID=abc123`) or saved to a `.env` file in this format:

```
UMAPI_CLIENT_ID=abc-123
UMAPI_CLIENT_SECRET=xyz-9876
UMAPI_ORG_ID=ABC123@AdobeOrg
```

Save this to a file with the name `.env` and the UMAPI CLI Tool will read it if
it is present in the current working directory.

> **NOTE:** You are responsible for keeping this file safe. Limit access to it
> to prevent unauthorized access.

For most users, it is sufficient to provide the Client ID, Client Secret and Org
ID. There are, however, cases where the auth or UMAPI endpoints need to be
customized. They can be set with these variables:

* `UMAPI_AUTH_HOST` - Hostname of auth endpoint server
* `UMAPI_AUTH_ENDPOINT` - Path portion of auth endpoint
* `UMAPI_URL` - Full URI to UMAPI endpoint

If you are working with more than one target, you can specify an alternative env
file with the `--env` option.

```
$ umapi --env .env-secondary user-read-all
```

The above example reads UMAPI config from the file `.env-secondary` since this
file isn't named `.env`, it won't be read automatically. To read it, we pass the
`--env` option before we specify the command (`user-read-all` in this case).

# Commands

## `user-read`

Get details for a single user by email address. 

Change the output format with the option `-f/--format`. Supported formats: JSON,
CSV, or human-readable (default.)

```
$ umapi user-read -e user@example.com
email     : user@example.com
groups    : ['All Apps', 'Adobe Stock']
username  : user@example.com
domain    : example.com
firstname : Example
lastname  : User
country   : US
type      : federatedID
```

Usage:

```
$ umapi user-read --help
Usage: umapi user-read [OPTIONS]

  Get details for a single user

Options:
  -h, --help                    Show this message and exit.
  -f, --format csv|json|pretty  Output format
  -e, --email TEXT              User email address  [required]
```

**Note:** The CSV format for `user-read` is the same as `user-read-all` ([see
below](#user-read-all)) with the exception of the `tags` column. This column can
currently only be read on a user-by-user basis and thus is not included in
`user-read-all` results.

## `user-read-all`

Get details for all users in a given console. 

Formats: [JSONL](http://jsonlines.org), CSV, or human-readable (default).

This command writes to stdout by default, but can optionally write output to a
given filename.

```
# write all users to a CSV file
$ umapi user-read-all -f csv -o users.csv
```

Usage:

```
$ umapi user-read-all --help
Usage: umapi user-read-all [OPTIONS]

  Get details for all users belonging to a console

Options:
  -h, --help                    Show this message and exit.
  -f, --format csv|json|pretty  Output format
  -o, --out-file FILENAME       Write output to this filename
  -g, --in-group GROUP          Limit query to members of GROUP
```

Names of columns/fields when writing to CSV or JSONL are the same.

| Column Name | Purpose                                                               |
|-------------|-----------------------------------------------------------------------|
| `type`      | User's identity type (`enterpriseID`, `federatedID` or `adobeID`)     |
| `firstname` | User's given (first) name                                             |
| `lastname`  | User's surname (last name)                                            |
| `email`     | User's email address                                                  |
| `username`  | SSO username of user                                                  |
| `domain`    | If the user is in a claimed domain, this is the domain they belong to |
| `country`   | Two-letter country code indicating country of user                    |
| `groups`    | List* of groups to which user is assigned                             |

\* in JSONL, groups are represented as a JSON list. In CSV, groups are
serialised to a comma-delimited list (enclosed in double quotes).

## `user-create`

Create a single user.

Example:

```
$ umapi user-create --type federatedID --email test.user.001@example.com \
  --username test.user.001@example.com --groups "All Apps,Adobe Sign" \
  --firstname Test --lastname "User 001" --country US
```

Usage:

```
$ umapi user-create --help
Usage: umapi user-create [OPTIONS]

  Create a single user.

Options:
  -h, --help                      Show this message and exit.
  --type adobeID|enterpriseID|federatedID
                                  User's identity type  [default: federatedID]
  --email TEXT                    User's email address  [required]
  --username TEXT                 User's username (set to email if omitted)
  --domain TEXT                   User's directory domain (set to username
                                  domain if omitted, if username is an email,
                                  this must be set to same domain)
  --groups TEXT                   Comma-delimited list of groups to assign
                                  user
  --firstname TEXT                User's first name
  --lastname TEXT                 User's last name
  --country TEXT                  User's two-letter (ISO-3166-1 alpha2)
                                  country code  [required]
```

## `user-create-bulk`

Create users in bulk from an input file.

Formats: [JSONL](http://jsonlines.org) or CSV (default)

Expects `-i/--in-file` option that specifies input file path.

Example - create all users specified in "users.csv"

```
$ umapi user-create-bulk -f csv -i users.csv
```

Usage:

```
$ umapi user-create-bulk --help
Usage: umapi user-create-bulk [OPTIONS]

  Create users in bulk from an input file

Options:
  -h, --help              Show this message and exit.
  -f, --format csv|json   Input file format  [default: csv]
  -i, --in-file FILENAME  Input filename
```

The columns/fields expected by `user-create-bulk` are named and formatted in the
same way as output from `user-read-all`.

| Column Name | Purpose                                                               |
|-------------|-----------------------------------------------------------------------|
| `type`      | User's identity type (`enterpriseID`, `federatedID` or `adobeID`)     |
| `firstname` | User's given (first) name                                             |
| `lastname`  | User's surname (last name)                                            |
| `email`     | User's email address                                                  |
| `username`  | SSO username of user                                                  |
| `domain`    | If the user is in a claimed domain, this is the domain they belong to |
| `country`   | Two-letter country code indicating country of user                    |
| `groups`    | List* of groups to which user is assigned                             |

\* in JSONL, groups are represented as a JSON list. In CSV, groups are
serialised to a comma-delimited list (enclosed in double quotes).

## `user-update`

Update a single user.

Example:

```
$ umapi user-update
  --email test.user.001@example.com \
  --firstname Test \
  --lastname "Username 001"
```

Usage:

```
$ umapi user-update --help
Usage: umapi user-update [OPTIONS]

  Update user information for a single user

Options:
  -h, --help                Show this message and exit.
  -e, --email TEXT          Email address that identifies the user  [required]
  -E, --email-new TEXT      Updated email address
  -f, --firstname TEXT      Updated given (first) name
  -l, --lastname TEXT       Updated surname (last name)
  -u, --username TEXT       Updated username
  -g, --groups-add TEXT     Comma-delimited list of groups to add for user
  -G, --groups-remove TEXT  Comma-delimited list of groups to remove from user
```

When updating a user, the email address is required to identify the user. The
`-e/--email` option specifies the identifying email address. `-u/--username` is
used to specify a new username when updating the username field.

Apart from `-e/--email`, all other options are not required.

## `user-update-bulk`

Update users in bulk from an input file.

Formats: [JSONL](http://jsonlines.org) or CSV (default)

Expects `-i/--in-file` option that specifies input file path.

Example - create all users specified in "users.csv"

```
$ umapi user-update-bulk -f csv -i users.csv
```

Usage:

```
$ umapi user-update-bulk --help
Usage: umapi user-update-bulk [OPTIONS]

  Update users in bulk from input file

Options:
  -h, --help              Show this message and exit.
  -f, --format csv|json   Input file format  [default: csv]
  -i, --in-file FILENAME  Input filename
```

This is the expected format of an input file for bulk updating users:

| Column Name     | Purpose                                        |
|-----------------|------------------------------------------------|

.changelog/changelog.md
| tag    | date       | title                 |
|--------|------------|-----------------------|
| v2.2.0 | 2023-07-21 | UMAPI CLI Tool v2.2.0 |

# New Features

* #7 Add option to `user-read-all` to limit users to members of specific group
* #8 Include `tags` field in `user-read-all` output

---

| tag    | date       | title                 |
|--------|------------|-----------------------|
| v2.1.0 | 2023-07-20 | UMAPI CLI Tool v2.1.0 |

# New Features

* #1 Include ID field in user-read and user-read-all
* #2 Infer output format in user/group read commands
* #3 Order CSV columns
* #4 Custom user-agent
* #5 Expose tags in user-read

---

| tag    | date       | title                 |
|--------|------------|-----------------------|
| v2.0.0 | 2023-07-11 | UMAPI CLI Tool v2.0.0 |

# Initial Stable Release

This is the initial stable release of the UMAPI CLI tool. For full
documentation, see `README.md`. If you run into any problems, please create an
issue on this repository.

---

| tag      | date       | title                   |
|----------|------------|-------------------------|
| v2.0.0a2 | 2023-06-29 | UMAPI CLI Tool v2.0.0a2 |

# Fixes

* Fix `.env` discovery for pipx installs

---

| tag      | date       | title                   |
|----------|------------|-------------------------|
| v2.0.0a1 | 2023-06-29 | UMAPI CLI Tool v2.0.0a1 |

Initial alpha release of the UMAPI CLI Tool.


umapi_cli/cli.py
# Copyright 2023 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.

import click
import sys
import io
import os
from umapi_client import UserQuery, UsersQuery, GroupsQuery
import dotenv
from pathlib import Path
from umapi_cli import config
from umapi_cli import client
from umapi_cli import formatter
from umapi_cli.action_queue import ActionQueue
from umapi_cli.formatter import normalize, InputHandler, OutputHandler, PassthroughHandler
from umapi_cli import log
from umapi_cli.version import __version__ as app_version

def _formatter(data_format, fh, handler):
    fmtr_class = getattr(formatter, data_format, None)
    if fmtr_class is None:
        click.echo("Unknown format '{}'".format(data_format))
        sys.exit(1)
    return fmtr_class(fh, handler)


def _output_fh(out_file=None):
    if out_file is not None:
        return open(out_file, 'w', encoding='utf-8')
    return sys.stdout


def _input_fh(in_file):
    return open(in_file, 'r')


def infer_format(filename):
    if filename.endswith('csv'):
        return 'csv'
    if filename.endswith('json') or filename.endswith('jsonl'):
        return 'json'
    return None


@click.group()
@click.option('--env', 'env_file', help="Path to .env file (optional)", default=None, type=click.Path())
@click.option('-t', '--test', 'test_mode', help="Run command in test mode", default=False, show_default=False,
              is_flag=True)
@click.option('-v', count=True, help="Enable verbose logging")
@click.help_option('-h', '--help')
@click.version_option(app_version, '--version', message='%(prog)s %(version)s')
@click.pass_context
def app(ctx, env_file, test_mode, v):
    log.init(v)
    if env_file is not None:
        dotenv.load_dotenv(env_file)
    else:
        env_file = dotenv.find_dotenv(usecwd=True)
        if env_file:
            dotenv.load_dotenv(env_file)
    ctx.ensure_object(dict)
    conf = config.get_options()
    ctx.obj['conn'] = client.create_conn(conf, test_mode)


def entry():
    debug = True if os.environ.get('UMAPI_DEBUG') == '1' else False
    try:
        app()
    except Exception as e:
        if not debug:
            click.echo(f"ERROR: {e}")
        else:
            raise e


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'output_format', help='Output format', metavar='csv|json|pretty', default='pretty',
              show_default=True)
@click.option('-e', '--email', help='User email address', required=True)
@click.pass_context
def user_read(ctx, output_format, email):
    """Get details for a single user"""

    fmtr = _formatter(output_format, _output_fh(), OutputHandler('user_read'))
    umapi_conn = ctx.obj['conn']
    user = UserQuery(umapi_conn, email).result()
    if not user:
        click.echo('No user found')
        sys.exit(1)
    fmtr.record(user)
    fmtr.write()


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'output_format', help='Output format', metavar='csv|json|pretty', show_default=True)
@click.option('-o', '--out-file', help='Write output to this filename', metavar='FILENAME')
@click.option('-g', '--in-group', help="Limit query to members of GROUP", metavar='GROUP')
@click.pass_context
def user_read_all(ctx, output_format, out_file, in_group):
    """Get details for all users belonging to a console"""

    if out_file is not None:
        output_format = infer_format(out_file)
    if output_format is None:
        output_format = 'pretty'

    fmtr = _formatter(output_format, _output_fh(out_file), OutputHandler('user_read_all'))
    umapi_conn = ctx.obj['conn']
    query = UsersQuery(umapi_conn, in_group=in_group)
    report_total = True
    for user in query:
        total, *_ = query.stats()
        if total is not None and report_total:
            log.info(f"Total records: {total}")
            report_total = False
        fmtr.record(user)
    fmtr.write()


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'output_format', help='Output format', metavar='csv|json|pretty', default='pretty',
              show_default=True)
@click.option('-g', '--group', 'group_name', help='Group name', required=True)
@click.pass_context
def group_read(ctx, output_format, group_name):
    """Get details for a single user group"""

    fmtr = _formatter(output_format, _output_fh(), OutputHandler('group_read'))
    umapi_conn = ctx.obj['conn']
    query = GroupsQuery(umapi_conn)
    matched = [g for g in query if normalize(g['groupName']) == normalize(group_name)]
    if len(matched) > 0:
        fmtr.record(matched[0])
        fmtr.write()
    else:
        click.echo(f"Group '{group_name}' not found")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'output_format', help='Output format', metavar='csv|json|pretty',
              show_default=True)
@click.option('-o', '--out-file', help='Write output to this filename', metavar='FILENAME')
@click.pass_context
def group_read_all(ctx, output_format, out_file):
    """Get details for all groups in a console"""

    if out_file is not None:
        output_format = infer_format(out_file)
    if output_format is None:
        output_format = 'pretty'

    fmtr = _formatter(output_format, _output_fh(out_file), OutputHandler('group_read'))
    umapi_conn = ctx.obj['conn']
    query = GroupsQuery(umapi_conn)
    for group in query:
        fmtr.record(group)
    fmtr.write()


@app.command()
@click.help_option('-h', '--help')
@click.option('--type', 'user_type', help="User's identity type", metavar='adobeID|enterpriseID|federatedID',
              default='federatedID', show_default=True)
@click.option('--email', help="User's email address", required=True)
@click.option('--username', help="User's username (set to email if omitted)")
@click.option('--domain', help="User's directory domain (set to username domain if omitted, if username is an "
                               "email, this must be set to same domain)")
@click.option('--groups', help="Comma-delimited list of groups to assign user", required=False)
@click.option('--firstname', help="User's first name")
@click.option('--lastname', help="User's last name")
@click.option('--country', help="User's two-letter (ISO-3166-1 alpha2) country code", required=True)
@click.pass_context
def user_create(ctx, user_type, email, username, domain, groups, firstname, lastname, country):
    """Create a single user."""
    """
       Example:

       umapi user-create \
             --type federatedID \
             --email test.user.001@example.com \
             --username test.username.001@example.com \
             --groups group1,group2 \
             --firstname Test \
             --lastname "User 001" \
             --country US
    """

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    if groups is None:
        groups = []
    else:
        groups = groups.split(',')
    queue.queue_user_create_action(user_type, email, country, firstname, lastname,
                                   username, domain, groups)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Create operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def user_create_bulk(ctx, input_format, in_file):
    """Create users in bulk from an input file"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('user_create_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for user in fmtr.read():
        if user['domain'] == '':
            user['domain'] = None
        queue.queue_user_create_action(id_type=user['type'],
                                       email=user['email'], username=user['username'],
                                       domain=user['domain'], groups=user['groups'],
                                       firstname=user['firstname'],
                                       lastname=user['lastname'],
                                       country=user['country'])
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


@app.command()
@click.help_option('-h', '--help')
@click.option('-e', '--email', help='User email address', required=True)
@click.option('-d', '--hard', 'hard_delete',
              help="Delete user from underlying directory instead of just the org level",
              default=False, show_default=False, is_flag=True)
@click.pass_context
def user_delete(ctx, email, hard_delete):
    """Delete a single user (from org and/or identity directory)"""

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    queue.queue_delete_action(email, hard_delete)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Delete operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def user_delete_bulk(ctx, input_format, in_file):
    """Delete users in bulk from input file (from org and/or identity directory)"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('user_delete_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for user in fmtr.read():
        queue.queue_delete_action(user['email'],
                                  True if user['hard_delete'] == 'y' else False)
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


@app.command()
@click.help_option('-h', '--help')
@click.option('-e', '--email', help='Email address that identifies the user', required=True)
@click.option('-E', '--email-new', help="Updated email address", required=False)
@click.option('-f', '--firstname', help="Updated given (first) name", required=False)
@click.option('-l', '--lastname', help="Updated surname (last name)", required=False)
@click.option('-u', '--username', help="Updated username", required=False)
@click.option('-g', '--groups-add', help="Comma-delimited list of groups to add for user", required=False)
@click.option('-G', '--groups-remove', help="Comma-delimited list of groups to remove from user", required=False)
@click.pass_context
def user_update(ctx, email, email_new, firstname, lastname, username, groups_add, groups_remove):
    """Update user information for a single user"""

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    if groups_add is not None:
        groups_add = groups_add.split(',')
    if groups_remove is not None:
        groups_remove = groups_remove.split(',')
    queue.queue_update_action(email, email_new=email_new, firstname=firstname,
                              lastname=lastname, username=username, add_groups=groups_add,
                              remove_groups=groups_remove)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Update operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def user_update_bulk(ctx, input_format, in_file):
    """Update users in bulk from input file"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('user_update_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for user in fmtr.read():
        queue.queue_update_action(**user)
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


@app.command()
@click.help_option('-h', '--help')
@click.option('--name', help="Name of group", required=True)
@click.option('--description', help="Optional group description")
@click.pass_context
def group_create(ctx, name, description):
    """Create a single user group."""
    """
       Example:

       umapi group-create \
             --name "Adobe Stock Users" \
             --description "Stock provisioning group"
    """

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    queue.queue_group_create_action(name, description)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Create operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def group_create_bulk(ctx, input_format, in_file):
    """Create groups in bulk from an input file"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('group_create_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for group in fmtr.read():
        queue.queue_group_create_action(group['name'], group['description'])
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


@app.command()
@click.help_option('-h', '--help')
@click.option('-n', '--name', help='Current name of group', required=True)
@click.option('-N', '--name-new', help="New name to assign group", required=False)
@click.option('-d', '--description', help="Updated group description", required=False)
@click.option('-u', '--users-add', help="Comma-delimited list of email addresses of users to assign", required=False)
@click.option('-U', '--users-remove', help="Comma-delimited list of email addresses of users to remove", required=False)
@click.option('-p', '--profiles-add', help="Comma-delimited list of product profiles to associate with group", required=False)
@click.option('-P', '--profiles-remove', help="Comma-delimited list of product profiles to remove from group", required=False)
@click.pass_context
def group_update(ctx, name, name_new, description, users_add, users_remove, profiles_add, profiles_remove):
    """Update information/memberships for a single group"""

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    if users_add is not None:
        users_add = users_add.split(',')
    if users_remove is not None:
        users_remove = users_remove.split(',')
    if profiles_add is not None:
        profiles_add = profiles_add.split(',')
    if profiles_remove is not None:
        profiles_remove = profiles_remove.split(',')
    queue.queue_group_update_action(name, name_new=name_new,
                                    description=description, add_users=users_add, remove_users=users_remove,
                                    add_profiles=profiles_add, remove_profiles=profiles_remove)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Update operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def group_update_bulk(ctx, input_format, in_file):
    """Update groups in bulk from input file"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('group_update_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for group in fmtr.read():
        queue.queue_group_update_action(**group)
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


@app.command()
@click.help_option('-h', '--help')
@click.option('-n', '--name', help='Group name', required=True)
@click.pass_context
def group_delete(ctx, name):
    """Delete a single user group"""

    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    queue.queue_group_delete_action(name)
    queue.execute()
    errors = queue.errors()
    if errors:
        click.echo("One or more errors occurred")
        click.echo(render_errors(errors).strip())
    else:
        click.echo("Delete operation succeeded")


@app.command()
@click.help_option('-h', '--help')
@click.option('-f', '--format', 'input_format', help='Input file format', metavar='csv|json', default='csv',
              show_default=True)
@click.option('-i', '--in-file', help='Input filename', metavar='FILENAME')
@click.pass_context
def group_delete_bulk(ctx, input_format, in_file):
    """Delete groups in bulk from input file"""

    fmtr = _formatter(input_format, _input_fh(in_file), InputHandler('group_delete_bulk'))
    umapi_conn = ctx.obj['conn']
    queue = ActionQueue(umapi_conn)
    for group in fmtr.read():
        queue.queue_group_delete_action(group['name'])
    completed = queue.execute()
    errors = queue.errors()
    print_bulk_summaries(completed, errors)


def render_errors(errors):
    i = 1
    error_str = []
    for messages in errors:
        error_str.append(f"Error {i} messages:")
        error_io = io.StringIO()
        fmtr = _formatter('pretty', error_io, PassthroughHandler())
        for message in messages:
            fmtr.record(message)
        fmtr.write()
        error_io.seek(0)
        for l in error_io.getvalue().split('\n'):
            error_str.append('   '+l)
        i += 1
    return '\n'.join(error_str)


def render_summary(summary):
    summary_io = io.StringIO()
    fmtr = _formatter('pretty', summary_io, PassthroughHandler())
    fmtr.record(summary)
    fmtr.write()
    summary_io.seek(0)
    return summary_io.getvalue()

def print_bulk_summaries(completed, errors):
    summary = {
        "Executed": completed,
        "Succeeded": completed-len(errors),
        "Errors": len(errors),
    }
    click.echo("--- Action Summary ---")
    click.echo(render_summary(summary).strip())
    click.echo("----------------------")
    if errors:
        click.echo("--- Error Summary ---")
        click.echo(render_errors(errors).strip())
        click.echo("--------------")


if __name__ == '__main__':
    entry()
